source/shared_patterns/numeric.rb

```diff
@@ -164,23 +164,28 @@ def numeric_constant(allow_user_defined_literals: false, separator:"'")
     # first a range (the whole number) is found
     # then, after the range is found, it starts to figure out what kind of number/constant it is
     # it does this by matching one of the includes
-    return PatternRange.new(
-        start_pattern: start_pattern,
-        end_pattern: end_pattern,
-        # only a single include pattern should match
+    return Pattern.new(
+        match: lookBehindToAvoid(/\w/).then(/\.?\d/).zeroOrMoreOf(valid_character),
         includes: [
-            # floating point
-            hex_prefix    .maybe(hex_digits    ).then(hex_point    ).maybe(hex_digits    ).maybe(hex_exponent    ).maybe(floating_suffix).then(hex_ending),
-            decimal_prefix.maybe(decimal_digits).then(decimal_point).maybe(decimal_digits).maybe(decimal_exponent).maybe(floating_suffix).then(decimal_ending),
-            # numeric
-            binary_prefix .then(binary_digits )                        .maybe(numeric_suffix).then(binary_ending ),
-            octal_prefix  .then(octal_digits  )                        .maybe(numeric_suffix).then(octal_ending  ),
-            hex_prefix    .then(hex_digits    ).maybe(hex_exponent    ).maybe(numeric_suffix).then(hex_ending    ),
-            decimal_prefix.then(decimal_digits).maybe(decimal_exponent).maybe(numeric_suffix).then(decimal_ending),
-            # invalid
-            newPattern(
-                match: oneOrMoreOf(valid_single_character.or(valid_after_exponent)),
-                tag_as: "invalid.illegal.constant.numeric"
+            PatternRange.new(
```

Comment:
Haha, very clever. I would imagine this helps maintain performance since the outside pattern is simple, and the complex inside patterns don't have to compete against all of the other patterns in an external context.

Reply:
It would have negligible performance impact as the original start_pattern was a lookbehind and a lookahead, and once the PatternRange was entered it would again no longer be competing with others.

Reply:
I agree that nesting helps performance, I just don't think there would be much difference between nesting with a begin/end pattern and a match/pattern.

Reply:
Yeah, I meant: the alternative solution of not using any ranges. Combining all of the internal cases into one gigantic or-separated pattern.

Reply:
(^ that was the first solution I had in mind for fixing #356)